File: src/api/chat/messageService.js
import api from '../axios'

const BASE_URL = '/api/chat'

/**
 * Gửi tin nhắn văn bản
 * @param {number} conversationId - ID hội thoại
 * @param {string} content - Nội dung tin nhắn
 * @returns {Promise<Object>} MessageDTO
 */
export const sendTextMessage = async (conversationId, content) => {
    const { data } = await api.post(`${BASE_URL}/conversations/${conversationId}/messages/text`, null, {
        params: { content }
    })
    return data
}

/**
 * Gửi tin nhắn emoji
 * @param {number} conversationId - ID hội thoại
 * @param {string} code - Mã emoji (ví dụ: "😀", "👍")
 * @returns {Promise<Object>} MessageDTO
 */
export const sendEmojiMessage = async (conversationId, code) => {
    const { data } = await api.post(`${BASE_URL}/conversations/${conversationId}/messages/emoji`, null, {
        params: { code }
    })
    return data
}

/**
 * Gửi tin nhắn đính kèm file
 * @param {number} conversationId - ID hội thoại
 * @param {string} messageText - Nội dung tin nhắn (optional)
 * @param {File[]} files - Danh sách file cần đính kèm
 * @returns {Promise<Object>} MessageDTO
 */
export const sendAttachmentMessage = async (conversationId, messageText, files) => {
    const formData = new FormData()
    if (messageText) {
        formData.append('messageText', messageText)
    }
    files.forEach(file => {
        formData.append('files', file)
    })

    const { data } = await api.post(
        `${BASE_URL}/conversations/${conversationId}/messages/attachments`,
        formData,
        {
            headers: {
                'Content-Type': 'multipart/form-data'
            }
        }
    )
    return data
}

/**
 * Thu hồi tin nhắn
 * @param {number} messageId - ID tin nhắn
 * @returns {Promise<Object>} MessageDTO
 */
export const recallMessage = async (messageId) => {
    const { data } = await api.post(`${BASE_URL}/messages/${messageId}/recall`)
    return data
}

/**
 * Xóa tin nhắn cho người dùng hiện tại
 * @param {number} messageId - ID tin nhắn
 * @returns {Promise<void>}
 */
export const deleteMessage = async (messageId) => {
    await api.delete(`${BASE_URL}/messages/${messageId}`)
}

/**
 * Đánh dấu tin nhắn đã đọc
 * @param {number} conversationId - ID hội thoại
 * @param {number} messageId - ID tin nhắn
 * @returns {Promise<void>}
 */
export const markMessageSeen = async (conversationId, messageId) => {
    await api.post(`${BASE_URL}/conversations/${conversationId}/messages/${messageId}/seen`)
}

File: be/src/main/java/com/giapho/coffee_shop_backend/chat/service/MessageService.java
package com.giapho.coffee_shop_backend.chat.service;

import com.giapho.coffee_shop_backend.chat.dto.MessageDTO;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface MessageService {

    /**
     * Gửi tin nhắn văn bản thuần tuý.
     */
    MessageDTO sendText(Long conversationId, String content);

    /**
     * Gửi tin nhắn emoji với mã emoji chuẩn (ví dụ "😀").
     */
    MessageDTO sendEmoji(Long conversationId, String emojiCode);

    /**
     * Gửi tin nhắn đính kèm tệp phương tiện (có thể kèm nội dung text).
     */
    MessageDTO sendAttachments(Long conversationId,
                               String messageText,
                               List<MultipartFile> files);

    /**
     * Thu hồi (recall) tin nhắn bởi người gửi.
     */
    MessageDTO recall(Long messageId);

    /**
     * Đánh dấu xoá tin nhắn ở phía người dùng hiện tại (không ảnh hưởng tới người khác).
     */
    void deleteForCurrentUser(Long messageId);

    /**
     * Đánh dấu đã xem tới một tin nhắn trong cuộc trò chuyện và trả về ID người dùng đã xem.
     */
    Long markSeen(Long conversationId, Long messageId);
}
